siriuspy/siriuspy/devices/fofb.py

```diff
@@ -0,0 +1,883 @@
+"""FOFB devices."""
+
+import time as _time
+import numpy as _np
+
+from mathphys.functions import get_namedtuple as _get_namedtuple
+
+from ..namesys import SiriusPVName as _PVName
+from ..search import BPMSearch as _BPMSearch, PSSearch as _PSSearch
+from ..fofb.csdev import HLFOFBConst as _Const
+
+from .device import Device as _Device, ProptyDevice as _ProptyDevice, \
+    Devices as _Devices
+from .bpm import BPMLogicalTrigger
+from .timing import Event
+from .pwrsupply import PowerSupplyFC
+from .psconv import StrengthConv
+
+NR_BPM = 160
+
+
+class _FOFBCtrlBase:
+    """FOFB Ctrl base."""
+
+    _devices = {
+        f'SI{i:02d}': f'IA-{i:02d}RaBPM:BS-FOFBCtrl' for i in range(1, 21)}
+    DEVICES = _get_namedtuple('DEVICES', *zip(*_devices.items()))
+
+
+class FOFBCtrlRef(_Device, _FOFBCtrlBase):
+    """FOFB controller device."""
+
+    _properties = (
+        'RefOrb-SP', 'RefOrb-RB',
+    )
+
+    def __init__(self, devname):
+        """Init."""
+        # check if device exists
+        if devname not in self.DEVICES:
+            raise NotImplementedError(devname)
+
+        # call base class constructor
+        super().__init__(devname, properties=FOFBCtrlRef._properties)
+
+    @property
+    def ref(self):
```

Comment:
@anacso17 , `reforbx`, `reforby`, `reforb` look more intuitive for me. are we using `ref`in other devices?

Reply:
We use `refx` and `refy`in SOFB device